source/components/temp_doeclim_component.cpp

```diff
@@ -0,0 +1,528 @@
+    //// Need to do this because if we're subject to a user constraint (being forced
+    //// to match a temperature record), need to track the Ftot that *would* have
+    //// produced the observed temperature record. This way there's a smooth
+    //// transition when we exit the constraint period, after which internal_Ftot
+    //// will rise in parallel with the value reported by ForcingComponent.
+    
+    // If we never had any temperature constraint, `internal_Ftot` will match `Ftot`.
+    //
+    // If the user has supplied temperature data, use that (except if past its end)
+    if( tgav_constrain.size() && runToDate <= tgav_constrain.lastdate() ) {
+    //    H_LOG( logger, Logger::NOTICE ) << "** Using user-supplied temperature" << std::endl;
+        H_LOG( logger, Logger::NOTICE ) << "** ERROR - DOECLIM can't currently handle user-supplied temperature" << std::endl;
+    }
+    //  tgav = tgav_constrain.get( runToDate );
+    //    tgaveq = tgav;
+    //
+    //    // Update last_Ftot and internal_Ftot for when we exit constraint
+    //    last_Ftot = core->sendMessage( M_GETDATA, D_RF_TOTAL ).value( U_W_M2 );
+    //    const double heatflux = core->sendMessage( M_GETDATA, D_HEAT_FLUX ).value( U_W_M2 );
+    //    internal_Ftot = 3.7 * tgav / S + heatflux;  // inverting normal T=S/3.7*F eqn
+    //} else {  // not constrained
+    //    // Not subject to a constraint. Get total forcing and update our
+    //    // internal tracking variables `internal_Ftot` (parallels Ftot, may
+    //    // be identical if never had a constraint) and `last_Ftot`.
+    
+
+	// A few forcing agents occur disproportionately over land, and thus exert
+	// a stronger effect than their global mean RF would suggest. In the future,
+	// we want to compute RF separately for land and ocean; for now, add a 'bonus'
+	// for these forcing agents.
+
+	// NOTE: this is commented out and not used for now.
+	/*
+	const double fbc = core->sendMessage( M_GETDATA, D_RF_BC ).value( U_W_M2 );
+	Ftot += fbc * bc_b; // % extra
+	const double foc = core->sendMessage( M_GETDATA, D_RF_OC ).value( U_W_M2 );
+	Ftot += foc * oc_b; // % extra
+	const double fso2d = core->sendMessage( M_GETDATA, D_RF_SO2d ).value( U_W_M2 );
+	Ftot += fso2d * so2d_b; // % extra
+	const double fso2i = core->sendMessage( M_GETDATA, D_RF_SO2i ).value( U_W_M2 );
+	Ftot += fso2i * so2i_b; // % extra
+	*/
+    
+    // Some needed inputs
+    int tstep = runToDate - core->getStartDate();
+    double aero_forcing = double(core->sendMessage( M_GETDATA, D_RF_BC ).value( U_W_M2 )) + double(core->sendMessage( M_GETDATA, D_RF_OC ).value( U_W_M2 )) + double(core->sendMessage( M_GETDATA, D_RF_SO2d ).value( U_W_M2 )) + double(core->sendMessage( M_GETDATA, D_RF_SO2i ).value( U_W_M2 ));
+    forcing[tstep] = double(core->sendMessage( M_GETDATA, D_RF_TOTAL ).value( U_W_M2 )) - ( 1.0 - alpha ) * aero_forcing;
+			    
+    // Initialize variables for time-stepping through the model
```

Comment:
Are the next 80 lines or so copied/translated from the original DOEClim code base?  If so, we should maybe include a pointer to that code, both so that anyone who wants to look at it can, and also to document that it was copied with the knowledge and permission of its authors and maintainers.